velox/functions/sparksql/tests/ConcatWsTest.cpp

```diff
@@ -0,0 +1,306 @@
+
+    std::string randomString;
+    for (std::size_t i = 0; i < length; ++i) {
+      randomString += chars[folly::Random::rand32() % chars.size()];
+    }
+    return randomString;
+  }
+
+  void testConcatWsFlatVector(
+      const std::vector<std::vector<std::string>>& inputTable,
+      const size_t argsCount,
+      const std::string& separator) {
+    std::vector<VectorPtr> inputVectors;
+
+    for (int i = 0; i < argsCount; i++) {
+      inputVectors.emplace_back(
+          BaseVector::create(VARCHAR(), inputTable.size(), execCtx_.pool()));
+    }
+
+    for (int row = 0; row < inputTable.size(); row++) {
+      for (int col = 0; col < argsCount; col++) {
+        std::static_pointer_cast<FlatVector<StringView>>(inputVectors[col])
+            ->set(row, StringView(inputTable[row][col]));
+      }
+    }
+
+    auto buildConcatQuery = [&]() {
+      std::string output = "concat_ws('" + separator + "'";
+
+      for (int i = 0; i < argsCount; i++) {
+        output += ",c" + std::to_string(i);
+      }
+      output += ")";
+      return output;
+    };
+    auto result = evaluate<FlatVector<StringView>>(
+        buildConcatQuery(), makeRowVector(inputVectors));
+
+    auto produceExpectedResult = [&](const std::vector<std::string>& inputs) {
+      auto isFirst = true;
+      std::string output;
+      for (int i = 0; i < inputs.size(); i++) {
+        auto value = inputs[i];
+        if (isFirst) {
+          isFirst = false;
+        } else {
+          output += separator;
+        }
+        output += value;
+      }
+      return output;
+    };
+
+    for (int i = 0; i < inputTable.size(); ++i) {
+      EXPECT_EQ(result->valueAt(i), produceExpectedResult(inputTable[i]))
+          << "at " << i;
+    }
+  }
+};
+
+TEST_F(ConcatWsTest, stringArgs) {
+  // Test with constant args.
+  auto rows = makeRowVector(makeRowType({VARCHAR(), VARCHAR()}), 10);
+  auto c0 = generateRandomString(20);
+  auto c1 = generateRandomString(20);
+  auto result = evaluate<SimpleVector<StringView>>(
+      fmt::format("concat_ws('-', '{}', '{}')", c0, c1), rows);
+  for (int i = 0; i < 10; ++i) {
+    EXPECT_EQ(result->valueAt(i), c0 + "-" + c1);
+  }
+
+  // Test with variable arguments.
+  size_t maxArgsCount = 10;
+  size_t rowCount = 100;
+  size_t maxStringLength = 100;
+
+  std::vector<std::vector<std::string>> inputTable;
+  for (int argsCount = 1; argsCount <= maxArgsCount; argsCount++) {
+    inputTable.clear();
+    inputTable.resize(rowCount, std::vector<std::string>(argsCount));
+
+    for (int row = 0; row < rowCount; row++) {
+      for (int col = 0; col < argsCount; col++) {
+        inputTable[row][col] =
+            generateRandomString(folly::Random::rand32() % maxStringLength);
+      }
+    }
+
+    SCOPED_TRACE(fmt::format("Number of arguments: {}", argsCount));
+    testConcatWsFlatVector(inputTable, argsCount, "--testSep--");
+    // Test with empty separator.
+    testConcatWsFlatVector(inputTable, argsCount, "");
```

Comment:
Please test the null separator

Reply:
> We may could register it as stateful VectorFunction but input argument is any like function hashWithSeed

This function doesn't support any input type. And as it is a bit special, we registered it as special form to allow it to be verified by fuzzer test.